specs/CustomDefaultDownloadLocation.md

```diff
@@ -0,0 +1,82 @@
+Custom Default Download Location
+===
+
+# Background
+This API allows you to set a custom default download location per profile. The
+user can still change the path through the Save As dialog.
+
+In this document we describe the updated API. We'd appreciate your feedback.
+
+# Examples
+```cpp
+HRESULT AppWindow::OnCreateCoreWebView2ControllerCompleted(
+    HRESULT result, ICoreWebView2Controller* controller)
+{
+    if (result == S_OK)
+    {
+        if (!m_webviewOption.downloadPath.empty())
+        {
+            Microsoft::WRL::ComPtr<ICoreWebView2Profile2> profile2;
+            CHECK_FAILURE(m_profile->QueryInterface(IID_PPV_ARGS(
+                &profile2)));
+            CHECK_FAILURE(profile2->put_DefaultDownloadPath(
+                m_webviewOption.downloadPath.c_str()));
+        }
+    }
+}
+```
+```c#
+void SetDefaultDownloadPathCmdExecuted(object target, ExecutedRoutedEventArgs e)
+{
+    var dialog = new TextInputDialog(
+        title: "Set Default Download Path",
+        description: "Enter the new default download path.",
+        defaultInput: WebViewProfile.DefaultDownloadPath);
+    if (dialog.ShowDialog() == true)
+    {
+        WebViewProfile.DefaultDownloadPath = dialog.Input.Text;
+    }
+}
+```
+# API Details
+```c#
+/// This is a continuation of the `ICoreWebView2Profile` interface.
+[uuid(DAF8B1F9-276D-410C-B481-58CBADF85C9C), object, pointer_default(unique)]
+interface ICoreWebView2Profile2 : ICoreWebView2Profile {
+  /// Gets the `DefaultDownloadPath` property. The default value is the
+  /// system default download path for the user.
+  [propget] HRESULT DefaultDownloadPath([out, retval] LPWSTR* value);
+
+  /// Sets the `DefaultDownloadPath` property. The value should be
+  /// an absolute path to a folder that the user and application can write to.
+  /// Returns `E_INVALIDARG` if the value is invalid, and the default download
+  /// path is not changed. If the directory does not yet exist, it is created
+  /// at the time of the next download. If the host application does not have
+  /// permission to create the directory, then the user is prompted to provide
+  /// a new path through the Save As dialog. This value is persisted in the user
+  /// data folder across sessions and can be changed by the user through the
```

Comment:
But there's no event to tell the host app that the user changed the default download path. They just have to poll for it.

Reply:
No event but they could check the path in the DownloadStarting event to see if it has changed

Reply:
When you say "This value ... can be changed by the user through the Save As dialog" do you mean the user can change the default download folder path through the save as dialog? If so, how do they do that - is it just whatever path the end user last selected becomes the new default download folder path?

Reply:
Yes whatever path they select becomes the new one

Reply:
And that's sort of the expected usage and that's why we don't have an event? We expect them to set an initial default download folder path and then the user can change it? As opposed to they set a default download folder path and that's the default every time?

Reply:
Yes exactly, they want to set the initial default path and allow users to change it

Reply:
Would be good to make this a little more explicit in the documentation and say that the default download folder path will change to wherever the end user last downloaded a file.

Given the scenario an event doesn't seem necessary.

Reply:
If I change the DefaultDownloadFolderPath inside the DownloadStarting event handler, will that change the default folder "just in time" for the upcoming download dialog?

Reply:
Sorry for the confusion. The user can override the default path through the Save As dialog per download, but this does not actually change the default download folder path. Only the Save dialog remembers the user's choice of location so that the next time the Save dialog is shown it suggests the user's last choice. Downloads that don't go through Save dialog still use the default download folder path.

Setting the DefaultDownloadFolderPath inside DownloadStarting event handler would be too late for the current download, and the ResultFilePath would also have to be updated.